File: frontend/src/logic/tasks/query_data.ts
import { ShowState } from "../models"
import { store, states } from "../store"
import { connection } from "../websocket"

export const query_data = (data: []) => {
    store.data = data
    states.data = ShowState.Output
}

export const query_data_send = () => {
    states.data = ShowState.Loading

    const msg = {
        type: "query_data",
        query: store.query,
    }

    connection.send(JSON.stringify(msg))
}

File: backend/tasks/query_data.py
import pandas as pd
import json
import storage


async def query_data(websocket, query):
    # Print colored Task name
    storage.printTask("query_data")

    # Establish new connection
    conn = storage.getEngine().connect()

    # Query the data
    df = pd.read_sql(query, con=conn)

    # Close the connection
    conn.close()

    # Store data as DataFrame in global storage
    storage.data = df

    # Transform DataFrame into a Dictonary
    # https://pandas.pydata.org/docs/reference/api/pandas.DataFrame.to_dict.html
    # [{'col1': 1, 'col2': 0.5}, {'col1': 2, 'col2': 0.75}]
    dataDict = df.head(300).to_dict("records")
    # dataDict = df.to_dict("records")

    # Response with data Dictornary attached
    response = {
        "task": "query_data",
        "data": dataDict
    }

    # Send Response
    await websocket.send(json.dumps(response))
